js/src/launchAsync.ts

```diff
@@ -63,6 +74,259 @@ let launchResponseResolve: LaunchResponseResolve;
+                try {
+                    options.onPreferencesChanged(e.data.substring(PostMessagePreferences.length));
+                } catch {
+                    // no-op?
+                }
+            }
+        }
+    };
+}
+
+function _makeSrcUrl(options: Options, funcType: FuncType): string {
+    let src = '';
+    if (funcType === FuncType.launchAsync) {
+        src = 'reader?exitCallback=ImmersiveReader-Exit';
+        if (options.cognitiveAppId) {
+            src += '&cognitiveAppId=' + options.cognitiveAppId;
+        }
+    } else if (funcType === FuncType.launchWithoutContentAsync) {
+        src = 'https://learningtools.onenote.com/learningtoolsapp/cognitive/reader?exitCallback=ImmersiveReader-Exit&skipearlygcm=true';
+    }
+
+    src += '&sdkPlatform=' + sdkPlatform + '&sdkVersion=' + sdkVersion;
+
+    src += '&cookiePolicy=' + ((options.cookiePolicy === CookiePolicy.Enable) ? 'enable' : 'disable');
+
+    if (options.hideExitButton) {
+        src += '&hideExitButton=true';
+    }
+
+    if (options.uiLang) {
+        src += '&omkt=' + options.uiLang;
+    }
+
+    return src;
+}
+
+function _generateIframeVariables(_parent: Node, useWebview: boolean, allowFullscreen: boolean): void {
+    parent = _parent ? _parent : document.body;
+    iframeContainer = document.createElement('div');
+    iframe = useWebview ? <HTMLIFrameElement>document.createElement('webview') : document.createElement('iframe');
+    iframe.allow = 'autoplay';
+    iframe.title = 'Immersive Reader Frame';
+    iframe.setAttribute('aria-modal', 'true');
+    if (allowFullscreen) { // Style iframe
+        iframe.setAttribute('allowfullscreen', '');
+    }
+    // Send an initial message to the webview so it has a reference to this parent window
+    if (useWebview) {
+        iframe.addEventListener('loadstop', () => {
+            iframe.contentWindow.postMessage(JSON.stringify({ messageType: 'WebviewHost' }), '*');
+        });
+    }
+    iframe.style.cssText = _parent ? 'position: static; width: 100%; height: 100%; left: 0; top: 0; border-width: 0' : 'position: static; width: 100vw; height: 100vh; left: 0; top: 0; border-width: 0';
+    noscroll = document.createElement('style');
+    noscroll.innerHTML = 'body{height:100%;overflow:hidden;}';
+}
+
+function _setIframeProps(src: string, optionParent: Node, uiZIndex: number): void {
+    iframe.src = src;
+
+    iframeContainer.style.cssText = optionParent ? `position: relative; width: 100%; height: 100%; border-width: 0; -webkit-perspective: 1px; z-index: ${uiZIndex}; background: white; overflow: hidden` : `position: fixed; width: 100vw; height: 100vh; left: 0; top: 0; border-width: 0; -webkit-perspective: 1px; z-index: ${uiZIndex}; background: white; overflow: hidden`;
+
+    iframeContainer.appendChild(iframe);
+    parent.appendChild(iframeContainer);
+
+    // Disable body scrolling
+    document.head.appendChild(noscroll);
+}
+
+function resetTimeout(): void {
+    if (timeoutId) {
```

Comment:
nit: You can safely call window.clearTimeout() with invalid values, no need for this check